apps/pipelines/nodes/history_middleware.py

```diff
@@ -0,0 +1,121 @@
+                # Since this response will get merged with the existing state messages, we cannot simply append the
+                # history to the user's message. We need to replace the full message history in the state.
+                # See https://github.com/langchain-ai/langchain/blob/c63f23d2339b2604edc9ae1d9f7faf7d6cc7dc78/libs/langchain_v1/langchain/agents/middleware/summarization.py#L286-L292
+                RemoveMessage(id=REMOVE_ALL_MESSAGES),
+                *self.node.get_history(self.session),
+                *state["messages"],
+            ]
+        }
+
+    def before_model(self, state, runtime):
+        result = super().before_model(state, runtime)
+        if result is not None:
+            # A result means that a summary was created
+            self.persist_summary(result["messages"])
+
+        return result
+
+    def persist_summary(self, messages: list[BaseMessage]):
+        checkpoint_message_id = self._find_latest_message_db_id(messages)
+        if not checkpoint_message_id:
+            logger.exception(
+                "Unable to persist summary: no checkpoint message ID found",
+                extra={"node_id": self.node.node_id, "history_mode": self.node.get_history_mode()},
+            )
+            return
+
+        # The first message is always a RemoveMessage if a summary was created
+        summary_message = messages[1]
+        self.node.store_compression_checkpoint(
+            compression_marker=summary_message.content, checkpoint_message_id=checkpoint_message_id
+        )
+
+    def _find_latest_message_db_id(self, messages: list) -> str | None:
+        """
+        Find the database ID of the latest message in the list that has one.
+        Typically it will be at index -2, since the one at -1 is the new user message.
+        """
+
+        for i in range(len(messages) - 2, -1, -1):
+            if "id" in messages[i].additional_kwargs:
+                return messages[i].additional_kwargs["id"]
+
+
+class SummarizeHistoryMiddleware(BaseNodeHistoryMiddleware):
+    """Summarizes overflowing history into a compact checkpoint."""
+
+    def __init__(self, *args, token_limit: int, **kwargs):
+        trigger = ("tokens", token_limit)
+        keep = ("messages", 20)
+        super().__init__(*args, trigger=trigger, keep=keep, **kwargs)
+
+
+class TruncateTokensHistoryMiddleware(BaseNodeHistoryMiddleware):
+    """Drops oldest messages whenever the token budget is exceeded."""
+
+    def __init__(self, *args, token_limit: int, **kwargs):
+        keep = trigger = ("tokens", token_limit)
+        super().__init__(*args, trigger=trigger, keep=keep, **kwargs)
+
+    def _create_summary(self, messages_to_summarize: list) -> str:
+        # Instead of creating a summary, we'll persist a compression marker. See _build_new_messages
+        return ""
+
+    def _build_new_messages(self, summary: str) -> list[HumanMessage]:
+        return [HumanMessage(content=COMPRESSION_MARKER)]
```

Comment:
won't this message get added to the pipeline state?

Reply:
Here we are: https://github.com/dimagi/open-chat-studio/pull/2560/commits/7034f087caa36b9a3cfbbf826a77900cc4ef4a1d and https://github.com/dimagi/open-chat-studio/pull/2560/commits/399402afcd9efa9e9035e706864091b43738e936